libcfnet/protocol.c

```diff
@@ -0,0 +1,90 @@
+/*
+   Copyright 2019 Northern.tech AS
+
+   This file is part of CFEngine 3 - written and maintained by Northern.tech AS.
+
+   This program is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by the
+   Free Software Foundation; version 3.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <protocol.h>
+
+#include <client_code.h>
+#include <client_protocol.h>
+#include <definitions.h>
+#include <net.h>
+#include <string_lib.h>
+#include <tls_generic.h>
+
+Seq *ProtocolOpenDir(AgentConnection *conn, const char *path)
+{
+    assert(conn != NULL);
+    assert(path != NULL);
+
+    char buf[CF_MSGSIZE] = {0};
+    int tosend = snprintf(buf, CF_MSGSIZE, "OPENDIR %s", path);
+    if (tosend < 0 || tosend >= CF_MSGSIZE)
```

Comment:
`tosend >= CF_MSGSIZE` is kind of redundant. `snprintf` can't exceed `CF_MSGSIZE` if it is limited by it.

Reply:
@karlhto Depends on the implementation. BSD man page:

```
     The snprintf() and vsnprintf() functions will write at most size-1 of
     the characters printed into the output string (the size'th character
     then gets the terminating `\0'); if the return value is greater than or
     equal to the size argument, the string was too short and some of the
     printed characters were discarded.
```

```
     These functions return the number of characters printed (not including
     the trailing `\0' used to end output to strings), except for snprintf()
     and vsnprintf(), which return the number of characters that would have
     been printed if the size were unlimited (again, not including the final
     `\0').  These functions return a negative value if an error occurs.
```

Reply:
Alright, thanks for the explanation @olehermanse :)
This should now be ready for merge IMO.